uswds/uswds
# USWDS - Header: Enable screen reader touch on mobile

## Summary
This PR makes the mobile header navigation items accessible to users navigating via touch screen readers. 

## Related issue
Closes #4787 

## Preview link
Preview link: [Storybook - Header component](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-header-touch/?path=/story/components-header--default)

## Problem statement
### 1. Mobile navigation is made inert when nested
When the mobile navigation is activated, all on-page elements (except the navigation and its ancestors) should be made inert by receiving the attributes `aria-hidden="true"` and `data-nav-hidden`. However,  these attributes are currently being applied only to direct children of `body`. This can make the header navigation inert if the page has any structure that nests the header _inside_ a direct child of body. 

The inert mobile navigation is most evident to users who navigate via touch screen readers (ex: VoiceOver Gestures on iOS and TalkBack on Android). For these users, the header nav items are completely invisible, which makes the site unnavigable. 

#### Problem Example
You can see the nesting issue in action in the current [Storybook header component](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/develop/?path=/story/components-header--default). Storybook inserts `div#root` into its page structure as a wrapper for all demo page content. The inert attributes are then applied to `div#root`, which makes all its nested USWDS page content inert, including the header. 

### 2. Mobile menu button is visible to users when menu is open
Additionally, the menu button should be hidden to all users when the menu is open. However, the current state allows the mobile menu button to remain visible to screen readers when the menu is open. 

## Solution
By carefully curating the selector for non-nav elements, we can:
1. Ensure that the header will remain visible to users navigating by touch screen readers
2. Hide the menu button to all users when the menu is open

## Testing and review
To test attributes in markup:
1. Open the [documentation demo page](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-header-touch/iframe.html?args=&id=pages-documentation-page--documentation-page&viewMode=story)
1. Narrow the screen to <1025px to active the mobile header structure
1. Inspect the page with web inspector
1. Open the flyout menu by clicking the menu button
1. Confirm that the `aria-hidden="true"`/`data-nav-hidden` attributes:
    1. are NOT applied to `.usa-header`, `.usa-nav-container`,  `.usa-nav`, or the parent of `.usa-header` (like div#root in Storybook)
    1. are applied to the header sibling elements like `usa-section`, `usa-footer`, and `usa-identifier`.

To test behavior in touch screen readers:  
1. Open the documentation page in iOS or android
2. Enable VoiceOver Gestures or TalkBack
3. When menu is open, confirm that:
    1. the menu items can be accessed via both swipe and tap
    2. the menu button can NOT be accessed via swipe or tap

Repeat tests for all header variants.

---

Before opening this PR, make sure you’ve done whichever of these applies to you:
